modules/distributor/tag_manager.go

```diff
@@ -0,0 +1,116 @@
+package distributor
+
+import (
+	"sync"
+	"time"
+
+	"github.com/grafana/tempo/modules/overrides"
+)
+
+const (
+	defaultUpdateCycle = time.Minute
+	tenantBuffer       = 20
+)
+
+type tagFilter func(tenant, tag string) bool
+
+type tagManager struct {
+	mtx sync.RWMutex
+	ch  chan string
+
+	// per tenant tag allow list
+	tenantAllowList map[string]map[string]struct{}
+
+	// globalTagsToDrop are a global list of tags to drop that apply to all tenants
+	globalTagsToDrop map[string]struct{}
+
+	o *overrides.Overrides
+}
+
+func NewTagManager(o *overrides.Overrides, globalTagDenyList []string) *tagManager {
+	// turn list into map for efficient checking
+	tagsToDrop := map[string]struct{}{}
+	for _, tag := range globalTagDenyList {
+		tagsToDrop[tag] = struct{}{}
+	}
+
+	tm := &tagManager{
+		ch:               make(chan string, tenantBuffer),
+		tenantAllowList:  map[string]map[string]struct{}{},
+		globalTagsToDrop: tagsToDrop,
+		o:                o,
+	}
+
+	go tm.updateTenantAllowList()
+
+	return tm
+}
+
+// todo: pass a ctx to exit gracefully
+func (tm *tagManager) updateTenantAllowList() {
+	ticker := time.NewTicker(defaultUpdateCycle)
+	for {
+		select {
+		case <-ticker.C:
+			// loop through all tenants
+			tm.mtx.RLock()
+			for tenant, currentMap := range tm.tenantAllowList {
+				newList := tm.o.SearchTagsAllowList(tenant)
+				match := true
+				for _, tag := range newList {
+					if _, ok := currentMap[tag]; !ok {
+						match = false
+					}
+				}
+				if len(currentMap) != len(newList) {
+					match = false
+				}
+				if match == false {
+					tm.ch <- tenant
+				}
+			}
+			tm.mtx.RUnlock()
+		case tenant := <-tm.ch:
+			// create the allowMap for this tenant
+			allowList := tm.o.SearchTagsAllowList(tenant)
+			allowMap := map[string]struct{}{}
+			for _, tag := range allowList {
+				allowMap[tag] = struct{}{}
+			}
+			tm.mtx.Lock()
+			tm.tenantAllowList[tenant] = allowMap
+			tm.mtx.Unlock()
+		}
+	}
+}
+
+// isAllowed is a function that decides whether a tag is extracted from traces into search data
+func (tm *tagManager) isAllowed(tenant, tag string) bool {
+	// first check per tenant allow list
+	tm.mtx.RLock()
+	defer tm.mtx.RUnlock()
+	if tenantAllowList, ok := tm.tenantAllowList[tenant]; ok {
+		if _, exists := tenantAllowList[tag]; exists {
+			// exists in tenant allow list, allow
+			return true
+		}
+	} else if allowList := tm.o.SearchTagsAllowList(tenant); len(allowList) > 0 { // fetch the allowList from overrides
+		// we want to take only a read lock here, pass it to the update loop
+		tm.ch <- tenant
```

Comment:
i don't understand what this channel is doing or why we are checking two allowlists for a tenant? is the first not sufficient?